### PR TITLE
Improvements to fetch, events, and url match

### PR DESCRIPTION
**Pull request by jaguillette**
Fetching paginated results no longer makes an extra call for the final page, instead determining the next page url from response headers. This also makes the urls in the fetch requests more sensible, since they don't have to include extra parameters so paginated fetch can add `&page=1` etc.

The time and date string constructors have been added to the Event object, which is now a class.

The script now checks the current page url for the `/courses/sis_course_id:123456` format. If it encounters that format, it redirects to the standard format and prompts the user to run the bookmarklet again. It also alerts the user if the script was unable to get any ID from the url.

Finally, the script now prompts the user to decide if they want to include descriptions for assignments and calendar events.

### CODE REVIEW

File: bookmarklet.js [hunk 1]
```diff
@@ -1,49 +1,94 @@
...
+        // since only one is needed, it's easier to extract with regex than a split
+        let parserExp = /<([\w:\/=?&.]+?)>; rel="next"/;
+        let nextLinkMatch = parserExp.exec(responseWithLinks.link);
+        if (nextLinkMatch) {
+          // if there's a next link, recurse this function
+          return paginatedFetch(nextLinkMatch[1], finalResponse);
+        } else {
+          // otherwise return the responses gathered
+          return finalResponse;
+        }
```
Comment (elliottyates) [inline bookmarklet.js:24]: Love the comments here.

File: bookmarklet.js [hunk 2]
```diff
@@ -144,5 +160,22 @@ void(
...
+    // and load it, while prompting the user to run the bookmarklet again after 
+    // the page loads.
+    alert(
+      "It looks like the URL for your Canvas page isn't quite what this tool requires. Your page should now reload with a URL that the tool can use. Once it loads back up, run this bookmark again to produce your syllabus."
+    );
+    const fetchCourse = fetch(`/api/v1/courses/${sisMatch[1]}`)
+      .then((res) => res.json())
+      .then((res) => {
+        window.location.href = `/courses/${res.id}`;
+      });
```
Comment (elliottyates) [inline bookmarklet.js:174]: FWIW the REST API also supports those `sis_course_id:123456`-style references. So I think you could just check for `match`, then set `match` = `sisMatch` if `match` doesn't exist, and use that in the same URL interpolations. I'm not sure if there's any other reason the script needs the original Canvas course ID vs the `sis_course_id` string, though. If it's only API calls, we probably don't need to redirect at all, just use that string as the course ID.

i.e. in postman, you can query `/api/v1/courses/sis_course_id:281762/activity_stream` or other course-related endpoints and get the same results as if you used the canvas ID.

**Review by elliottyates — APPROVED** [review]
Awesome!

File: bookmarklet.js (hunk 2 above)
Comment (jaguillette) [reply bookmarklet.js:174]: I did notice that, but unfortunately the `sis_course_id` syntax doesn't work as a parameter for the calendar endpoint: `/api/v1/calendar_events?context_codes[]=course_${courseId}&all_events=true` won't work if `courseId` is in the sis format. I didn't see a way to use the SIS ID as a context code, and didn't want to get too far into making a seamless solution for this edge case, so a redirect is where I landed.

Comment (elliottyates) [reply bookmarklet.js:174]: Cool! I didn't know that, gtk. Thanks!

**Pull request merged**